src/lib/scroller.js

```diff
@@ -40,32 +17,63 @@ function Scroller(consoleElement) {
   //pre-bind functions and throttle expansion
   this.refresh = this._renderVisible.bind(this);
   this.scroll = this._onScroll.bind(this);
-  this.expand = _.throttle(this._expand.bind(this), 150, {
+  this.expandTop = _.throttle(this._expandTop.bind(this), 150, {
+    leading: true,
+    trailing: true
+  });
+  this.expandBottom = _.throttle(this._expandBottom.bind(this), 150, {
     leading: true,
     trailing: true
   });
 }
 
-Scroller.prototype.setLines = function(newLines) {
-  var len = newLines.length;
+Scroller.prototype._generateContent = function(){
+  return _(this.lines)
+    .slice(this.startPosition - this.lineOffset, this.endPosition - this.lineOffset)
+    .map(function(line){
+      if(line.length === 0){
+        // insert a blank space to prevent pre omitting a trailing newline,
+        // even though pre/pre-nowrap/pre-line are specified.
+        return '\u2009';
+      }
+      return line;
+    })
+    .join('\n');
+};
+
+Scroller.prototype.setLines = function(newLines, offset) {
   this.lines = newLines;
+  this.lineOffset = offset;
   if(this.sticky){
-    this.startPosition = Math.max(0, len - this.minVisible);
-  }else if(len === 1 && newLines[0].length === 0){
+    this.startPosition = Math.max(this.lineOffset, this.lineCount() - this.visibleCount);
+    this.endPosition = this.lineCount();
```

Comment:
define this above `this.startPosition` so you don't have to call `this.lineCount()` twice and then use `this.endPosition` in the `this.startPosition` calculation.